oci/oci.go

```diff
@@ -0,0 +1,256 @@
+// Package oci provides basic operations for manipulating OCI images.
+package oci
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/opencontainers/go-digest"
+	"github.com/opencontainers/image-spec/specs-go"
+	spec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// Init initializes the img directory as an OCI image.
+// i.e. Creates oci-layout, index.json, and blobs.
+//
+// img directory must not exist before calling this function.
+//
+// imageLayoutVersion can be an empty string for specifying the default version.
+func Init(img, imageLayoutVersion string) error {
+	if imageLayoutVersion == "" {
+		imageLayoutVersion = spec.ImageLayoutVersion
+	}
+	if _, err := os.Stat(img); err == nil {
+		return os.ErrExist
+	}
+	// Create the directory
+	if err := os.MkdirAll(img, 0755); err != nil {
+		return err
+	}
+	// Create blobs/sha256
+	if err := os.MkdirAll(
+		filepath.Join(img, "blobs", string(digest.Canonical)),
+		0755); err != nil {
+		return nil
+	}
+	// Create oci-layout
+	if err := WriteImageLayout(img, spec.ImageLayout{Version: imageLayoutVersion}); err != nil {
+		return err
+	}
+	// Create index.json
+	return WriteIndex(img, spec.Index{Versioned: specs.Versioned{SchemaVersion: 2}})
+}
+
+func blobPath(img string, d digest.Digest) string {
+	return filepath.Join(img, "blobs", d.Algorithm().String(), d.Hex())
+}
+
+func indexPath(img string) string {
+	return filepath.Join(img, "index.json")
+}
+
+// BlobReader reads an OCI blob.
+type BlobReader interface {
```

Comment:
In general, I would always just return an `io.ReadCloser` and then allow detection of seekability. This keeps implementations from being bound by seek, which isn't always necessary. You can see the seek detection code in `content.WriteBlob` for an example.